fix(policychecks): Update Policy Checks to use profile selected by AWS Toolkits instead of always default

kevluu-aws opened the pull request:
## Problem
Policy Checks currently only uses the default profile within the credentials file for the CLI tool and SDK calls made to A2.

## Solution
Pass on the profile name from the current profile setup in AWS Toolkits to the CLI tool and SDK.

---

License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

jpinkney-aws reviewed (approved):
When someone from your team approves I'll merge it

mponaws commented on packages/core/src/awsService/accessanalyzer/vue/iamPolicyChecks.ts:
```diff
@@ -179,7 +179,9 @@ export class IamPolicyChecksWebview extends VueWebview {
                             documentType,
```
Firstly, thank you @kevluu-aws for fixing it! 🙌

mponaws commented on packages/core/src/awsService/accessanalyzer/vue/iamPolicyChecks.ts:
```diff
@@ -919,6 +943,11 @@ export function isJsonPolicyLanguage(document: string) {
 
+export function getProfileName(): string | undefined {
+    // We neeed to split the name on 'profile:' to extract the correct profile name
+    return globals.awsContext.getCredentialProfileName()?.split('profile:')[1]
```
Is this a constant format? I am trying to understand who enforces this format?

mponaws replied:
Also, can we add an error message on the backend to determine any incorrect format here?

kevluu-aws replied:
Toolkits uses the format in displaying the profile name to users after authenticating through the extension. It's stored with `profile:` prepended in the Credentials object.

mponaws commented on packages/core/src/test/awsService/accessanalyzer/iamPolicyChecks.test.ts:
```diff
@@ -155,6 +155,8 @@ describe('validatePolicy', function () {
                 'us-east-1',
```
Can we add a test that will help us determine if there is any change to the expected format in the `getProfileName()` function?

kevluu-aws replied:
The CLI tool tests check for a profile to be passed as `undefined`. This will check if the 'profile:' part gets changed

mponaws replied:
Had a chat with Kevin offline regarding if there is a way for us to catch the change in the format of the profile name, and he confirmed that it should be caught in the unit tests.

justinmk3 commented:
Does this fix a user-facing symptom? If so, please add a changelog as described in https://github.com/aws/aws-toolkit-vscode/pull/6074#issuecomment-2489829941

kevluu-aws commented:
Added changelog describing the fix

justinmk3 commented:
Please try doing a squash-merge locally (`git merge --squash`, or whatever tool you prefer). That will fix the commit history here.

If this still adds 1700 lines of code then next question is why :)

kevluu-aws commented:
> Please try doing a squash-merge locally (`git merge --squash`, or whatever tool you prefer). That will fix the commit history here.
> 
> If this still adds 1700 lines of code then next question is why :)

Accidental rebasing caused it, just fixed it and rebased with updated changelog

justinmk3 commented:
The jscpd ci check (for redundant code) is failing, and its logs point to code that could easily be deduplicated with minimal effort.

justinmk3 commented on packages/core/src/test/awsService/accessanalyzer/iamPolicyChecks.test.ts:
```diff
@@ -579,6 +589,8 @@ describe('customChecks', function () {
                     '--region',
                     'us-east-1',
+                    '--profile',
+                    'undefined',
```
All of the tests assert "undefined". Is this PR intended to support the non-undefined case?

kevluu-aws replied:
These test cases do not actually ever use this argument, as we stub the actual execution of our CLI tool here. It simply looks at the global `globals.awsContext.getCredentialProfileName()`, which is `undefined` in the test execution context.

Is there a way to change the current profile in the test execution context? If we are able to, we can assert it is propagated properly.

justinmk3 replied:
> Is there a way to change the current profile in the test execution context?

Look at existing code.

justinmk3 commented:
Merging to unblock. Please followup on the above comments.
